core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/waiters/AsyncWaiterHandler.java

```diff
@@ -0,0 +1,118 @@
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * Handler for async waiter operations
+ * @param <T> the type of the response
+ */
+@SdkInternalApi
+@NotThreadSafe
+public final class AsyncWaiterHandler<T> {
+    private static final Logger log = Logger.loggerFor(AsyncWaiterHandler.class);
+    private final PollingStrategy pollingStrategy;
+    private final ScheduledExecutorService executorService;
+    private final WaiterHandlerHelper<T> handlerHelper;
+    private int attemptNumber = 0;
+
+    public AsyncWaiterHandler(PollingStrategy pollingStrategy,
+                              List<WaiterAcceptor<T>> waiterAcceptors,
+                              ScheduledExecutorService executorService) {
+        this.pollingStrategy = Validate.paramNotNull(pollingStrategy, "pollingStrategy");
+        Validate.paramNotNull(waiterAcceptors, "waiterAcceptors");
+        this.executorService = Validate.paramNotNull(executorService, "executorService");
+        this.handlerHelper = new WaiterHandlerHelper<>(waiterAcceptors, pollingStrategy);
+    }
+
+    /**
+     * Execute the provided async polling function
+     */
+    CompletableFuture<WaiterResponse<T>> execute(Supplier<CompletableFuture<T>> asyncPollingFunction) {
+        log.info(() -> "starting to execute");
+        CompletableFuture<WaiterResponse<T>> future = new CompletableFuture<>();
+        doExecute(asyncPollingFunction, future);
+        return future;
+    }
+
+    private void doExecute(Supplier<CompletableFuture<T>> asyncPollingFunction, CompletableFuture<WaiterResponse<T>> future) {
+        ++attemptNumber;
+        log.info(() -> "runOnce" + attemptNumber);
+        runAsyncPollingFunction(asyncPollingFunction, future);
+    }
+
+    private void runAsyncPollingFunction(Supplier<CompletableFuture<T>> asyncPollingFunction,
+                                         CompletableFuture<WaiterResponse<T>> future) {
+        asyncPollingFunction.get().whenComplete((response, exception) -> {
+            Optional<WaiterState> waiterState = handlerHelper.nextWaiterStateIfMatched(response, exception);
+
+            if (waiterState.isPresent()) {
+                log.info(() -> "waiter state " + waiterState.get());
+                WaiterState state = waiterState.get();
+                switch (state) {
+                    case SUCCESS:
+                        future.complete(handlerHelper.createWaiterResponse(response, exception, attemptNumber));
+                        break;
+                    case RETRY:
+                        maybeRetry(asyncPollingFunction, future);
+                        break;
+                    case FAILURE:
+                        future.completeExceptionally(SdkClientException.create("A waiter acceptor was matched and transitioned "
+                                                                               + "the waiter to failure state"));
```

Comment:
Any way that we could let an acceptor can provide a failure reason to include in the exception?

Reply:
Yeah, let me try it